README.rst
===============================
Draft in a Flask
===============================

.. image:: https://badge.fury.io/py/draftin-a-flask.png
    :target: http://badge.fury.io/py/draftin-a-flask
    
.. image:: https://travis-ci.org/waynew/draftin-a-flask.png?branch=master
        :target: https://travis-ci.org/waynew/draftin-a-flask

.. image:: https://pypip.in/d/draftin-a-flask/badge.png
        :target: https://crate.io/packages/draftin-a-flask?version=latest


A simple Flask server that allows you to publish Pelican blags from 
http://draftin.com using
[WebHooks](https://draftin.com/documents/69898?token=5fjKKlZ0-AeBzqj_RAftAGdzRzl9VBfBHj5wpSWm_gU)

* Free software: BSD license
* Documentation: http://draftin-a-flask.rtfd.org.

Usage
-----

::

    $ pip install draftin_a_flask
    $ env DIF_CONTENT=/path/to/content \
    DIF_OUTPUT=/path/to/output \
    DIF_PELICAN=/path/to/pelican_binary draftican
    Listening at endpoint QRFky1tR0KqHGM3cJoitwEi8tTpknaNnMpNHHiTIm8
    * Running on http://0.0.0.0:5678/
    * Restarting with reloader
    Listening at endpoint QRFky1tR0KqHGM3cJoitwEi8tTpknaNnMpNHHiTIm8
    
(Yes, it displays the print twice. I'm sure there's a way to get around it. If
it bothers you, I accept pull requests!)

Setup your WebHook from within Draft, and now you can write your blog posts in
Draft and easily publish.


Future Features
---------------

* Automatic uploads using rsync/ssh/file copy
* Settings provided in a file
* Improved error handling (e.g. missing title, etc.)


Known Bugs
----------

* If you're missing important fields (like title and date) it probably will
  skip publishing that doc.


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2014-01-13)
++++++++++++++++++

* First release on PyPI.

tests/test_draftin-a-flask.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_draftin-a-flask
----------------------------------

Tests for `draftin-a-flask` module.
"""

import unittest

from draftin-a-flask import draftin-a-flask


class TestDraftin-a-flask(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()

tests/test_draftin_a_flask.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_draftin_a_flask
----------------------------------

Tests for `draftin_a_flask` module.
"""

import pytest
import os
import json
import shutil

try:
    from imp import reload
except ImportError:
    pass

from mock import MagicMock, call

from draftin_a_flask import draftin_a_flask


@pytest.fixture
def secret_file():
    package_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    secret_file = os.path.join(package_root, 'draftin_a_flask', 's3kret.key')
    draftin_a_flask.setup()
    return secret_file


def test_if_secret_keyfile_is_missing_it_should_create_on_setup(secret_file):
    if os.path.isfile(secret_file):
        os.unlink(secret_file)
    assert not os.path.exists(secret_file)

    draftin_a_flask.setup()

    assert os.path.isfile(secret_file)


def test_after_delete_secret_file_should_contain_different_string(secret_file):
    if os.path.isfile(secret_file):
        os.unlink(secret_file)
    draftin_a_flask.setup()
    with open(secret_file) as f:
        old_secret = f.read()
    os.unlink(secret_file)

    draftin_a_flask.setup()

    with open(secret_file) as f:
        new_secret = f.read()
    assert old_secret != new_secret


def test_GET_to_secret_endpoint_should_405(secret_file):
    reload(draftin_a_flask)
    test_client = draftin_a_flask.app.test_client()
    with open(secret_file) as f:
        key = f.read()
    rv = test_client.get(key)
    assert rv.status_code == 405


def test_POST_to_secret_endpoint_should_call_publish():
    reload(draftin_a_flask)
    test_client = draftin_a_flask.app.test_client()
    draftin_a_flask.publish = MagicMock()
    data = {'id': 12345,
            'name': "Some document",
            'content': 'Some markdown',
            'content_html': 'Some html',
            #'user': {'id': 42, 'email': 'fnord@fnord'},
            "created_at": "2013-05-23T14:11:54-05:00",
            "updated_at": "2013-05-23T14:11:58-05:00"
            }
    rv = test_client.post(draftin_a_flask.app.secret_key,
            data={'payload':json.dumps(data)})
            
    expected = [call.publish(data['name'], data['content'])]
    assert draftin_a_flask.publish.mock_calls == expected
    

def test_POST_to_secret_endpoint_with_no_content_should_500():
    reload(draftin_a_flask)
    test_client = draftin_a_flask.app.test_client()
    data = {'id': 12345,
            'name': "Some document",
            'content_html': 'Some html',
            'user': {'id': 42, 'email': 'fnord@fnord'},
            "created_at": "2013-05-23T14:11:54-05:00",
            "updated_at": "2013-05-23T14:11:58-05:00"
            }
    rv = test_client.post(draftin_a_flask.app.secret_key,
            data=json.dumps(data),
                          content_type='application/json')
    assert rv.status_code == 500


def test_if_content_directory_does_not_exist_it_should_get_made():
    draftin_a_flask.subprocess = MagicMock()
    content = os.path.join(draftin_a_flask.ROOT,
                           draftin_a_flask.CONTENT)
    if os.path.exists(content):
        shutil.rmtree(content)

    draftin_a_flask.publish('whatev', 'This is some *content*')

    is_dir = os.path.isdir(content)
    assert is_dir


def test_passing_name_and_content_to_publish_should_write_it_to_content_dir():
    draftin_a_flask.subprocess = MagicMock()
    draftin_a_flask.CONTENT = 'content'
    name = "fnordy fnord"
    fname = os.path.join(draftin_a_flask.ROOT,
                         draftin_a_flask.CONTENT,
                         name.replace(' ', '-') + '.md')

    draftin_a_flask.publish(name, 'This is some *content*')

    is_file = os.path.isfile(fname)
    assert is_file


def test_publish_should_actually_write_content():
    draftin_a_flask.subprocess = MagicMock()
    draftin_a_flask.CONTENT = 'fnord'
    expected_content = "This is some sweet *content*"
    name = "fnordy fnord"
    fname = os.path.join(draftin_a_flask.ROOT,
                         draftin_a_flask.CONTENT,
                         name.replace(' ', '-') + '.md')

    draftin_a_flask.publish(name, expected_content)

    with open(fname) as f:
        actual_content = f.read()
        assert actual_content == expected_content

    shutil.rmtree(os.path.join(draftin_a_flask.ROOT, draftin_a_flask.CONTENT))


def test_passing_content_to_publish_should_call_pelican():
    draftin_a_flask.subprocess = MagicMock()
    expected = [call.check_output([draftin_a_flask.PELICAN,
                                   draftin_a_flask.CONTENT, 
                                   '-o', 
                                   draftin_a_flask.OUTPUT,
                                   '-s',
                                   draftin_a_flask.PELICANCONF]),
                ]

    draftin_a_flask.publish('a name', 'This is some content')

    assert draftin_a_flask.subprocess.mock_calls == expected


def test_if_environment_values_are_set_they_should_be_preferred():
    pelicanconf = 'repelicant'
    pelican = 'pelican or pelicant?'
    content = "Well isn't that special?"
    output = 'Buuuuuurrrrp' # get it?
    os.environ['DIF_PELICANCONF'] = pelicanconf
    os.environ['DIF_PELICAN'] = pelican
    os.environ['DIF_CONTENT'] = content
    os.environ['DIF_OUTPUT'] = output

    reload(draftin_a_flask)

    assert draftin_a_flask.PELICANCONF == pelicanconf
    assert draftin_a_flask.PELICAN == pelican
    assert draftin_a_flask.CONTENT == content
    assert draftin_a_flask.OUTPUT == output


def test_if_draftican_file_is_present_it_should_read_conf_values():
    with open('.draftican', 'w') as f:
        f.write(json.dumps(dict(OUTPUT='output',
                                CONTENT='content',
                                PELICAN='pelican',
                                PELICANCONF='pelicanconf')))

    reload(draftin_a_flask)
    os.unlink('.draftican')

    assert draftin_a_flask.PELICANCONF == 'pelicanconf'
    assert draftin_a_flask.PELICAN == 'pelican'
    assert draftin_a_flask.CONTENT == 'content'
    assert draftin_a_flask.OUTPUT == 'output'


draftin_a_flask/draftin_a_flask.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import
import os
import re
import json
import subprocess
from . import utils
from flask import Flask, request

app = Flask(__name__)
if os.path.isfile('.draftican'):
    with open('.draftican') as f:
        data = json.load(f)
        os.environ['DIF_OUTPUT'] = data.get('OUTPUT')
        os.environ['DIF_CONTENT'] = data.get('CONTENT')
        os.environ['DIF_PELICAN'] = data.get('PELICAN')
        os.environ['DIF_PELICANCONF'] = data.get('PELICANCONF')

OUTPUT = os.environ.get('DIF_OUTPUT', 'path/to/output')
CONTENT = os.environ.get('DIF_CONTENT', 'path/to/input')
PELICAN = os.environ.get('DIF_PELICAN', '/path/to/pelican')
PELICANCONF = os.environ.get('DIF_PELICANCONF', '/path/to/pelicanconf.py')
ROOT = os.path.dirname(os.path.abspath(__file__))
SECRET_FILE = os.path.join(ROOT, 's3kret.key')


def setup():
    if not os.path.isfile(SECRET_FILE):
        with open(SECRET_FILE, 'w') as f:
            f.write(''.join(utils.random_string(size=42)))

setup()
with open(SECRET_FILE) as f:
    app.secret_key = f.read()


@app.route('/'+app.secret_key, methods=['POST'])
def main():
    data = json.loads(request.form.get('payload'))
    name = data.get('name')
    content = data.get('content')
    #if content is None:
    #    return 'Missing content', 500
    publish(name, content)
    return "OK"


def publish(name, content):
    name = re.sub(r'[^a-zA-Z\\/]', '-', name) + '.md'
    dir_ = os.path.join(ROOT, CONTENT)
    if not os.path.exists(dir_):
        os.makedirs(dir_)
    fname = os.path.join(dir_, name)
    with open(fname, 'w') as f:
        f.write(content)
    output = subprocess.check_output([PELICAN,
                                      CONTENT, 
                                      '-o', 
                                      OUTPUT,
                                      '-s',
                                      PELICANCONF])
    app.logger.debug(output)


if __name__ == "__main__":
    app.run('0.0.0.0', port=5678, debug=True)
